### PR TITLE
feat: add keepkey settings & initialize flow

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

Adds KeepKey settings to the V2 app, as specified in:

1. https://shapeshift.notion.site/KeepKey-Settings-86cce69e059346c0b14aa0547703465a
2. https://shapeshift.notion.site/Initialize-Recover-KeepKey-on-V2-15129a0c73fc4ace995191552721db69

### Out of scope for this PR, but will come as a fast-follow:

- Ability to recover an existing KeepKey wallet
- PIN modal included inside the settings menu

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

1. Closes https://github.com/shapeshift/web/issues/628
2. Closes https://github.com/shapeshift/web/issues/1300

## Risk

Minimal - mostly self-contained, with some additions to `walletProvider.tsx`.

## Testing

Note, the functionality is behind a feature flag: `REACT_APP_KEEP_KEY_SETTINGS`

Connect a KeepKey to the app and use the new KeepKey menu to change the device's settings, wipe the device, and re-initialize it.

## Screenshots (if applicable)

Screen Shot 2022-04-06 at 3 22 38 pm

Screen Shot 2022-04-06 at 3 23 09 pm

### CODE REVIEW

File: src/context/WalletProvider/KeepKey/hooks/useKeepKeyEventHandler.ts [hunk 1]
```diff
@@ -95,23 +105,39 @@ export const useKeepKeyEventHandler = (state: KeyringState, dispatch: Dispatch<A
...
+      })
+    }
+
+    const updateStatus = (status: UpdateStatus) => {
+      dispatchKeepKey({
+        type: KeepKeyActions.SET_UPDATE_STATUS,
+        payload: status
+      })
+    }
+
```
Comment (gomesalexandre) [inline src/context/WalletProvider/KeepKey/hooks/useKeepKeyEventHandler.ts:128]: I'd move these two action creators in `<KeepKeyProvider />`, next to the `reset` action creator:
https://github.com/shapeshift/web/pull/1341/commits/3a6c512aab2eaf53db09f2eb8deab9da5bfb8ce2#diff-81fd9ebe2a1638c6ea22ea8a13af9bbc5476682929f3cfb42a4f8e1c1aa11391R50

Comment (0xApotheosis) [reply src/context/WalletProvider/KeepKey/hooks/useKeepKeyEventHandler.ts]: Fantastic suggestion, thank you ser: https://github.com/shapeshift/web/pull/1341/commits/244c943fc559b7d1cbe97b2b7f1fd64b77275e3f

File: src/context/WalletProvider/KeepKey/utils.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { HDWallet } from '@shapeshiftoss/hdwallet-core'
+import { KeepKeyHDWallet } from '@shapeshiftoss/hdwallet-keepkey'
+
+export const isKeepKeyWallet = (wallet: HDWallet | null): wallet is KeepKeyHDWallet => {
+  return (wallet as KeepKeyHDWallet)?._isKeepKey
```
Comment (gomesalexandre) [inline src/context/WalletProvider/KeepKey/utils.ts:5]: Can't we use `isKeepKey()` instead? That's effectively doing the exact same type guard logic we're doing in `hdwallet`: https://github.com/shapeshift/hdwallet/blob/70bef5b63bd2feaac308ead8e4aa202ba79cfbe4/packages/hdwallet-keepkey/src/keepkey.ts

File: src/context/WalletProvider/KeepKeyProvider.tsx [hunk 3]
```diff
@@ -48,6 +55,21 @@ const KeepKeyContext = createContext<IKeepKeyContext | null>(null)
 
 export const KeepKeyProvider = ({ children }: { children: React.ReactNode }): JSX.Element => {
   const [state, dispatch] = useReducer(reducer, initialState)
+  const { state: walletState, dispatch: walletDispatch, load } = useWallet()
+  const { wallet } = walletState
+  const keepKeyWallet = isKeepKeyWallet(wallet) ? wallet : undefined
```
Comment (gomesalexandre) [inline src/context/WalletProvider/KeepKeyProvider.tsx:60]: Can we do this check only twice? e.g once when there is no `wallet` defined yet, a second time after `wallet` is defined. `useMemo()` would allow us to compute the keepKeyWallet variable in a memoized fashion if we use `wallet` as a dependency

**Review by gomesalexandre — COMMENTED** [review]
Added two small comments regarding accessing wallet / wallet type discrimination but overall logic looks solid!

File: src/context/WalletProvider/KeepKey/utils.ts (hunk 2 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/KeepKey/utils.ts]: Right you are: https://github.com/shapeshift/web/pull/1341/commits/b35046c7403c58ee9436cda3e8c8d9be2a688ba7

File: src/context/WalletProvider/KeepKeyProvider.tsx (hunk 3 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/KeepKeyProvider.tsx]: Very nice, ser: https://github.com/shapeshift/web/pull/1341/commits/b35046c7403c58ee9436cda3e8c8d9be2a688ba7

**Comment by 0xdef1cafe** [comment]
image

@DiggyDiggy2 @SS-FOX-McCloud @reallybeard i reckon we should be significantly more aggressive with the copy here, if this does what i think it does (wipes mnemonic from device) users will definitely lose access to funds. if this doesn't wipe the mnemonic we should still change the copy.

if it wipes the mnemonic:

"I understand that by wiping my KeepKey I will lose access to my funds if I don't have my recovery phrase.

[ ] I have backed up my recovery phrase and understand what I am doing"

need to click the checkbox to enable the button

if it doesn't wipe the mnemonic:

change title to "Reset to default settings"

change copy to "Reset your KeepKey to factory default settings. Please note this will not wipe your seed phrase" etc

thoughts? 🤷‍♂️

**Comment by 0xdef1cafe** [comment]
image

back button needs a teeny bit of padding

**Comment by SS-FOX-McCloud** [comment]
@0xdef1cafe I agree! It will wipe the mnemonic so the more aggressive copy would work great!
 "I understand that by wiping my KeepKey I will lose access to my funds if I don't have my recovery phrase.
> 
> [ ] I have backed up my recovery phrase and understand what I am doing"
> 
> need to click the checkbox to enable the button

**Review by theoboldfrazier — APPROVED** [review]
This all looks really good. I pulled down and tested it out and everything was working well.

**Comment by gomesalexandre** [comment]
The timeout options would probably need to be full width of their container as opposed to currently, that padding + background-color combination looks off:
image

Expected:

image

To match exactly the mockups however, the selected option should be without a background-color, only the hovered option should have it (quick fiddle, that's not exactly the hover color but DevTools rugged me on hovered state):

image

File: src/components/Layout/Header/NavBar/KeepKey/LastDeviceInteractionStatus.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  const translate = useTranslate()
+  const {
+    state: { lastDeviceInteractionStatus }
+  } = useWallet()
+  const greenShade = useColorModeValue('green.700', 'green.200')
+  const yellowShade = useColorModeValue('yellow.500', 'yellow.200')
+
+  return lastDeviceInteractionStatus ? (
+    <Alert
+      status={lastDeviceInteractionStatus === 'success' ? 'success' : 'error'}
```
Comment (gomesalexandre) [inline src/components/Layout/Header/NavBar/KeepKey/LastDeviceInteractionStatus.tsx:20]: Nitpick: If we make `Outcome` to be `export type Outcome = 'success' | 'error' | undefined`, you could directly use it here instead of a ternary, since you won't reach this branch in case it's undefined.

**Review by gomesalexandre — APPROVED** [review]
LGTM! Tested and it all works great. Added one small UI nitpick that's non-blocking and can be tackled as a follow-up commit/PR as a ☕ time.

File: src/components/Layout/Header/NavBar/KeepKey/LastDeviceInteractionStatus.tsx (hunk 4 above)
Comment (0xApotheosis) [reply src/components/Layout/Header/NavBar/KeepKey/LastDeviceInteractionStatus.tsx]: I like it - updated in https://github.com/shapeshift/web/pull/1341/commits/284f1ac95af39579135508ea712026483fb6cbc1

**Comment by 0xApotheosis** [comment]
Made the wipe modal more aggressively worded.

Screen Shot 2022-04-08 at 11 44 39 am

**Comment by 0xApotheosis** [comment]
> The timeout options would probably need to be full width of their container as opposed to currently, that padding + background-color combination looks off: image

This should be all sorted now @gomesalexandre 👌

**Comment by gomesalexandre** [comment]
> > The timeout options would probably need to be full width of their container as opposed to currently, that padding + background-color combination looks off
> 
> This should be all sorted now @gomesalexandre 👌

Tested again and it looks perfect! Well done Ser 🍬

**Review by gomesalexandre — APPROVED** [review]
Another one for that great work 🥇

**Comment by 0xApotheosis** [comment]
> Another one for that great work 🥇 

Aw shucks @gomesalexandre - I'll have to get trophy shelf.

File: headers.js [hunk 5]
```diff
@@ -6,6 +6,8 @@ const cspMeta = Object.entries({
   'connect-src': [
     "'self'",
     'data:',
+    // Explicitly whitelist our KeepKey versions file
+    'https://bafybeied24gc2ipvlxdbs4v676dwho2l5aafmngrleic3do2czdvgb546u.ipfs.dweb.link/keepKey.json',
```
Comment (cjthompson) [inline headers.js:10]: where did this URL come from?

please document in a comment details about where to get more information about how to find/generate this URL.

Why are we not just pulling the firmware list directly from GitHub?

File: sample.env [hunk 6]
```diff
@@ -19,6 +19,7 @@ REACT_APP_FEATURE_PLUGIN_BITCOIN=false
 REACT_APP_FEATURE_PLUGIN_COSMOS=false
 REACT_APP_FEATURE_GEM_RAMP=false
 REACT_APP_REDUX_LOGGING=false
+REACT_APP_KEEP_KEY_SETTINGS=false
```
Comment (cjthompson) [inline sample.env:22]:
```suggestion
REACT_APP_FEATURE_KEEPKEY_SETTINGS=false
```

File: src/AppProviders.tsx [hunk 7]
```diff
@@ -33,15 +34,17 @@ export function AppProviders({ children }: ProvidersProps) {
...
-                    <ModalProvider>
-                      <PortfolioProvider>
-                        <MarketDataProvider>
-                          <TransactionsProvider>
-                            <DefiManagerProvider>{children}</DefiManagerProvider>
-                          </TransactionsProvider>
-                        </MarketDataProvider>
-                      </PortfolioProvider>
-                    </ModalProvider>
+                    <KeepKeyProvider>
```
Comment (cjthompson) [inline src/AppProviders.tsx:37]: Another provider?

File: src/components/Layout/Header/NavBar/ExpandedMenuItem.tsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    <MenuItem
+      display='flex'
+      alignItems='center'
+      px={3}
+      {...props}
+      closeOnSelect={!hasSubmenu}
+      isDisabled={isDisabled}
+      style={isDisabled ? disabledStyleOverride : undefined}
+    >
+      <RawText flex={1}>{label}</RawText>
```
Comment (cjthompson) [inline src/components/Layout/Header/NavBar/ExpandedMenuItem.tsx:64]: I'd prefer that this be a `<Text>` component and that `label` is a string pointing to a translation. Is there a need for `label` to be dynamic text and not translated?

File: src/components/Layout/Header/NavBar/KeepKey/AwaitKeepKey.tsx [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  return awaitingDeviceInteraction ? (
+    <>
+      <Divider />
+      <Box p={3}>
+        <Flex {...props}>
+          <InfoIcon color={blueShade} mt={1} />
+          <Box ml={3}>
+            <Text translation={translation} mb={3} fontWeight='medium' color={blueShade} />
+            <Button colorScheme='blue' variant='ghost-filled' onClick={cancel} size='sm'>
+              {translate('common.cancel')}
```
Comment (cjthompson) [inline src/components/Layout/Header/NavBar/KeepKey/AwaitKeepKey.tsx:43]: does `Button` not take a `t` prop to pass in a translation path string?

File: src/components/Layout/Header/NavBar/KeepKey/ChangeLabel.tsx [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  const { state } = useWallet()
+  const { walletInfo } = state
+  const { keepKeyWallet } = useKeepKey()
+  const {
+    state: { awaitingDeviceInteraction }
+  } = useWallet()
+  const [keepKeyLabel, setKeepKeyLabel] = useState(walletInfo?.name)
+
+  const handleChangeLabelInitializeEvent = async () => {
+    await keepKeyWallet?.applySettings({ label: keepKeyLabel })
```
Comment (cjthompson) [inline src/components/Layout/Header/NavBar/KeepKey/ChangeLabel.tsx:24]: possible `UnhandledPromiseRejection` error

File: src/components/Layout/Header/NavBar/KeepKey/ChangePassphrase.tsx [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    state: { hasPassphrase }
+  } = useKeepKey()
+  const {
+    state: { awaitingDeviceInteraction }
+  } = useWallet()
+
+  const handleToggle = async () => {
+    const currentValue = !!hasPassphrase
+    setHasPassphrase(!hasPassphrase)
+    await keepKeyWallet?.applySettings({ usePassphrase: !currentValue })
```
Comment (cjthompson) [inline src/components/Layout/Header/NavBar/KeepKey/ChangePassphrase.tsx:26]: possible `UnhandledPromiseRejection` error

File: src/context/WalletProvider/KeepKey/hooks/useKeepKeyVersions.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  firmware: VersionStatus
+}
+
+export const useKeepKeyVersions = () => {
+  const [versions, setVersions] = useState<Versions>()
+  const { state } = useWallet()
+  const { wallet } = state
+
+  const getBootloaderVersion = (keepKey: KeepKeyHDWallet, releases: FirmwareReleases): string => {
+    const hash = keepKey.features?.bootloaderHash.toString() ?? ''
```
Comment (cjthompson) [inline src/context/WalletProvider/KeepKey/hooks/useKeepKeyVersions.ts:41]: `wallet.features` only gets populated after calling `getFeatures()` (async). That may automatically happen in the `initialize` call but be aware that the property is asynchronous.

File: src/context/WalletProvider/KeepKey/hooks/useKeepKeyVersions.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    const hex = buffer.toString('hex')
+    return releases.hashes.bootloader[hex.toLowerCase()]
+  }
+
+  useEffect(() => {
+    if (!wallet || !isKeepKey(wallet)) return
+    ;(async () => {
+      const { data: releases } = await axios.get<FirmwareReleases>(
+        // We'll need to update this whenever new firmware/bootloader is released.
+        'https://bafybeied24gc2ipvlxdbs4v676dwho2l5aafmngrleic3do2czdvgb546u.ipfs.dweb.link/keepKey.json',
```
Comment (cjthompson) [inline src/context/WalletProvider/KeepKey/hooks/useKeepKeyVersions.ts:52]: You should put this into `config.ts`.

File: src/context/WalletProvider/KeepKeyProvider.tsx [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+export enum DeviceTimeout {
+  TenMinutes = '600000',
+  FifteenMinutes = '900000',
+  TwentyMinutes = '1200000',
+  ThirtyMinutes = '1800000',
+  FortyFiveMinutes = '2700000',
+  SixtyMinutes = '3600000'
+}
+
+export const timeoutOptions: RadioOption<DeviceTimeout>[] = [
```
Comment (cjthompson) [inline src/context/WalletProvider/KeepKeyProvider.tsx:22]:
```suggestion
export const timeoutOptions: RadioOption<DeviceTimeout>[] = Object.freeze([
```

File: src/context/WalletProvider/KeepKeyProvider.tsx [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+  const [state, dispatch] = useReducer(reducer, initialState)
+
+  const setHasPinCaching = useCallback((payload: boolean | undefined) => {
+    dispatch({
+      type: KeepKeyActions.SET_HAS_PIN_CACHING,
+      payload
+    })
+  }, [])
+
+  const setHasPassphrase = useCallback((payload: boolean | undefined) => {
```
Comment (cjthompson) [inline src/context/WalletProvider/KeepKeyProvider.tsx:114]: what benefit are you getting from `useCallback`?

File: src/context/WalletProvider/WalletViewsSwitch.tsx [hunk 16]
```diff
@@ -28,13 +29,18 @@ export const WalletViewsSwitch = () => {
...
     history.goBack()
     // If we're back at the select wallet modal, remove the initial route
     // otherwise clicking the button for the same wallet doesn't do anything
     if (history.location.pathname === '/') {
       dispatch({ type: WalletActions.SET_INITIAL_ROUTE, payload: '' })
     }
+
+    // If we are interacting with a KeepKey, cancel any active requests on back
+    if (state.wallet && isKeepKey(state.wallet)) {
+      await state.wallet.cancel()
```
Comment (cjthompson) [inline src/context/WalletProvider/WalletViewsSwitch.tsx:42]: possible `UnhandledPromiseRejection` error

**Comment by cjthompson** [comment]
Has this been tested when unplugging/plugging in the keepkey while the settings modals are open? We need to gracefully handle disconnections.

File: src/state/slices/preferencesSlice/preferencesSlice.ts [hunk 17]
```diff
@@ -23,7 +24,8 @@ const initialState: Preferences = {
     CosmosPlugin: getConfig().REACT_APP_FEATURE_PLUGIN_COSMOS,
     GemRamp: getConfig().REACT_APP_FEATURE_GEM_RAMP,
     FoxyInvestor: getConfig().REACT_APP_FEATURE_FOXY_INVESTOR,
-    ReduxLogging: getConfig().REACT_APP_REDUX_LOGGING
+    ReduxLogging: getConfig().REACT_APP_REDUX_LOGGING,
+    KeepKeySettings: getConfig().REACT_APP_KEEP_KEY_SETTINGS
```
Comment (cjthompson) [inline src/state/slices/preferencesSlice/preferencesSlice.ts:28]:
```suggestion
    KeepKeySettings: getConfig().REACT_APP_FEATURE_KEEPKEY_SETTINGS
```

Comment (0xApotheosis) [reply src/state/slices/preferencesSlice/preferencesSlice.ts]: Renamed in https://github.com/shapeshift/web/pull/1341/commits/4ff5e2e7b910ed1038a6e6b9e6aca478245fb2af.

File: sample.env (hunk 6 above)
Comment (0xApotheosis) [reply sample.env]: Renamed in https://github.com/shapeshift/web/pull/1341/commits/4ff5e2e7b910ed1038a6e6b9e6aca478245fb2af.

File: src/components/Layout/Header/NavBar/ExpandedMenuItem.tsx (hunk 8 above)
Comment (0xApotheosis) [reply src/components/Layout/Header/NavBar/ExpandedMenuItem.tsx]: It was only to keep `ExpandedMenuItem` extensible, but I can also see value in forcing the caller to have defined translations.

Updated in https://github.com/shapeshift/web/pull/1341/commits/08b4791f902eaf6700816c015425629378f59bc2.

File: src/components/Layout/Header/NavBar/KeepKey/AwaitKeepKey.tsx (hunk 9 above)
Comment (0xApotheosis) [reply src/components/Layout/Header/NavBar/KeepKey/AwaitKeepKey.tsx:43]: It doesn't, but I do see value in adding our own button wrapper that handles this, much like we've already done for `Text`.

File: src/context/WalletProvider/KeepKey/hooks/useKeepKeyVersions.ts (hunk 13 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/KeepKey/hooks/useKeepKeyVersions.ts]: Added as config in https://github.com/shapeshift/web/pull/1341/commits/47f1a601be3bc9b889045bc85bbbbdef4f9b43e7.

File: src/context/WalletProvider/WalletViewsSwitch.tsx (hunk 16 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/WalletViewsSwitch.tsx]: Handled in https://github.com/shapeshift/web/pull/1341/commits/f9fe9b56dee0d189360272b2cc4ae6e9a2d20f34

File: src/components/Layout/Header/NavBar/KeepKey/ChangePassphrase.tsx (hunk 11 above)
Comment (0xApotheosis) [reply src/components/Layout/Header/NavBar/KeepKey/ChangePassphrase.tsx]: Handled in https://github.com/shapeshift/web/pull/1341/commits/f9fe9b56dee0d189360272b2cc4ae6e9a2d20f34

File: src/components/Layout/Header/NavBar/KeepKey/ChangeLabel.tsx (hunk 10 above)
Comment (0xApotheosis) [reply src/components/Layout/Header/NavBar/KeepKey/ChangeLabel.tsx]: Handled in https://github.com/shapeshift/web/pull/1341/commits/f9fe9b56dee0d189360272b2cc4ae6e9a2d20f34

File: src/context/WalletProvider/KeepKeyProvider.tsx (hunk 15 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/KeepKeyProvider.tsx:114]: The `useEffect` below depends on `setHasPassphrase` (as well as `setDeviceTimeout` and `setHasPinCaching`).

If these functions aren't wrapped by `useCallback` they'll be recreated every render, creating an infinite render loop (or at least exceeding the render stack depth limit).

TL;DR, without it the app crashes and we get a white screen.

File: src/context/WalletProvider/KeepKeyProvider.tsx (hunk 14 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/KeepKeyProvider.tsx]: Addressed in https://github.com/shapeshift/web/pull/1341/commits/543178a225b970f55f09b7736d44c9209f70c03d

File: src/context/WalletProvider/KeepKey/hooks/useKeepKeyVersions.ts (hunk 12 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/KeepKey/hooks/useKeepKeyVersions.ts:48]: Got it - I added a check and call in https://github.com/shapeshift/web/pull/1341/commits/543178a225b970f55f09b7736d44c9209f70c03d to mitigate this.

File: src/AppProviders.tsx (hunk 7 above)
Comment (0xApotheosis) [reply src/AppProviders.tsx:37]: I've gone back and forth on this a few times in my head.

I've landed on `KeepKeyProvider` containing too much hardware wallet-specific logic to muddy `WalletProvider` with. I expect if we add first-class support for another wallet we'll refactor `KeepKeyProvider` into a `HardwareWallet` provider.

**Comment by 0xApotheosis** [comment]
> Has this been tested when unplugging/plugging in the keepkey while the settings modals are open? We need to gracefully handle disconnections.

Yep, it works smoothly - it reconnects and fetches the latest device data.

**Review by cjthompson — APPROVED** [review]
